src/main/java/com/iexec/core/replicate/ReplicatesService.java

```diff
@@ -303,9 +304,62 @@ public void updateReplicateStatus(String chainTaskId,
         }
 
         if (!canUpdate) {
+            log.warn("Cannot update replicate {}",
+                    getStatusUpdateLogs(chainTaskId, replicate, statusUpdate));
+            return Optional.of(ReplicateStatusUpdateError.GENERIC_CANT_UPDATE);
+        }
+
+        return Optional.empty();
+    }
+
+    public Optional<TaskNotificationType> updateReplicateStatus(String chainTaskId,
+                                                                String walletAddress,
+                                                                ReplicateStatusUpdate statusUpdate) {
+        return updateReplicateStatus(chainTaskId, walletAddress, statusUpdate, false);
+    }
+
+    /*
+     * We retry up to 100 times in case the task has been modified between
+     * reading and writing it.
+     *
+     * Before updating we check:
+     *   1) if valid transition.
+     *   2) if worker did fail when CONTRIBUTE/REVEAL/UPLOAD_FAILED.
+     *   3) if worker did succeed onChain when CONTRIBUTED/REVEALED.
+     *   4) if worker did upload when RESULT_UPLOADING.
+     */
+    @Retryable(value = {OptimisticLockingFailureException.class}, maxAttempts = 100)
+    public Optional<TaskNotificationType> updateReplicateStatus(String chainTaskId,
+                                                                String walletAddress,
+                                                                ReplicateStatusUpdate statusUpdate,
+                                                                boolean skipUpdateAbilityTests) {
+        log.info("Replicate update request [status:{}, chainTaskId:{}, walletAddress:{}, details:{}]",
+                statusUpdate.getStatus(), chainTaskId, walletAddress, statusUpdate.getDetailsWithoutStdout());
+
+        if (!skipUpdateAbilityTests
+                && canUpdateReplicateStatus(chainTaskId,walletAddress,statusUpdate).isPresent()) {
             return Optional.empty();
         }
 
+        ReplicatesList replicatesList = getReplicatesList(chainTaskId).orElseThrow();           // "get" could be used there but triggers a warning
+        Replicate replicate = replicatesList.getReplicateOfWorker(walletAddress).orElseThrow(); // "get" could be used there but triggers a warning
+        ReplicateStatus newStatus = statusUpdate.getStatus();
+
+        if (newStatus.equals(CONTRIBUTED)) {
+            int workerWeight = iexecHubService.getWorkerWeight(walletAddress);
```

Comment:
I think I see some code duplication with canUpdateReplicateWeight

Reply:
Code duplication has been deleted in the new version. It should be better now! :)
By the way, args are not checked until they are needed. I don't think it is an issue there.